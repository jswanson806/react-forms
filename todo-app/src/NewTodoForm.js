import React, { useState } from 'react';

const NewTodoForm = ({ addTodo }) => {
    const INITIAL_STATE = {
        task: ''
    }
    const [formData, setFormData] = useState(INITIAL_STATE);

    const handleSubmit = (e) => {
        e.preventDefault();
        addTodo(formData.task);
        setFormData(INITIAL_STATE);
    }

    const handleChange = (e) => {
        const {name, value} = e.target;
        setFormData(formData => ({
            ...formData,
            [name]: value
        }));
    }

    return (
        <form onSubmit={handleSubmit}>
            <label htmlFor="task">Task: </label>
            <input 
                id="task"
                name="task"
                placeholder="task"
                value={formData.task}
                onChange={handleChange}
            />
            <button>Add Todo</button>
        </form>
    )
}

export default NewTodoForm;