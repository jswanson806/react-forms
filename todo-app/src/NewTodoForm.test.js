import React from "react";
import { render, fireEvent } from "@testing-library/react";
import NewTodoForm from './NewTodoForm';

it("should render without crashing", () => {
    render(<NewTodoForm />)
})

it("should match the snapshot", () => {
    const { asFragment } = render(<NewTodoForm />);
    expect(asFragment()).toMatchSnapshot();
})

it("should handle user input correctly", () => {
    const { getByLabelText } = render(<NewTodoForm />);
    const task = getByLabelText('Task:');

    fireEvent.change(task, { target: { value: 'dishes' } });
    // check that new input value is "dishes"
    expect(task).toHaveValue("dishes");
})

it("should submit with correct data", () => {
    // spy on addTodo
    const addTodo = jest.fn();
    const { getByRole, getByLabelText } = render(<NewTodoForm addTodo={addTodo}/>);
    const task = getByLabelText('Task:');
    const btn = getByRole('button', {name: "Add Todo"})

    fireEvent.change(task, { target: { value: 'dishes' } });
    fireEvent.click(btn);

    // check that addTodo was called only once
    expect(addTodo).toHaveBeenCalledTimes(1);
    // check that addTodo was called with "dishes"
    expect(addTodo).toHaveBeenCalledWith("dishes");
})