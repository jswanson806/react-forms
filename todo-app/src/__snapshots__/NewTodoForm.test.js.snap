// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot 1`] = `
<DocumentFragment>
  <form>
    <label
      for="task"
    >
      Task: 
    </label>
    <input
      id="task"
      name="task"
      placeholder="task"
      value=""
    />
    <button>
      Add Todo
    </button>
  </form>
</DocumentFragment>
`;
